vim: Add global marks

AidanV opened the pull request:
Closes https://github.com/zed-industries/zed/issues/13111

Release Notes:

- vim: Added global marks `'[A-Z]`
- vim: Added persistence for global (and local) marks. When re-opening the same workspace your previous marks will be available.
- vim: Added sentence marks `'(` and `')`

ConradIrwin committed:
vim-global

AidanV committed:
Local Mark Persistince WIP

AidanV committed:
global marks opens file

AidanV committed:
Switching files correctly but no jump

AidanV committed:
Database not writable

AidanV committed:
fixed warnings

AidanV committed:
working non-anchored global/local marks

AidanV committed:
Anchored marks

AidanV committed:
Global mark database kind of working

AidanV committed:
Switched to using MarksCollection

AidanV committed:
Fixed loading anchors

AidanV committed:
harpoon and other marks

AidanV committed:
Got rid of warnings

AidanV committed:
Removed some comments

AidanV committed:
clippy...

AidanV committed:
Jumping correctly when opening buffer

AidanV committed:
cleanup

AidanV committed:
switching to buffers with no file in other panes

AidanV committed:
Switched to text Anchors and handling file changes

AidanV committed:
fixing marks for tests

AidanV committed:
fixing windows OsString

AidanV committed:
more tests passing

AidanV committed:
Fixed "test_command_visual_replace"

AidanV commented:
@ConradIrwin I am having trouble fixing this last test [here](https://github.com/AidanV/zed/blob/95c90d763c3cc9b477ba6bd7df1d26d0bbc98450/crates/vim/src/test.rs#L1088C1-L1088C6). I am not sure why it is failing since local marks work as expected. Some help would be greatly appreciated.

ConradIrwin commented:
:+1: Will take a look

ConradIrwin committed:
Use normal mark flow for "special" marks

They should be persisted like normal

ConradIrwin committed:
TEMP

ConradIrwin commented:
@AidanV I took a pass at this today, though there's a bunch more to do.

In particular:
* I removed special marks, we should treat these as any normal marks
* I added support for multi buffer marks (that are not persisted) in addition to singleton buffer marks
* I added a bit of fine-grained logic around the database updates (though I haven't wired them up to the actual DB yet)
* I tried to refactor things to simplify the APIs (there's now just get_mark/set_mark that most callers can use) and the response can be matched differently.

Still todo:
* We need the marks collection to work even if the workspace doesn't have a database_id(). I was going to change the key of the marks state to be an entity id of the workspace instead. We will only serialize to disk if the database_id() exists when we need it.
* Actually wire up the disk updates
* Test the changes I made
* Merge main

I plan to continue working on this a bit over the next few days, but please let me know if you're going to pick things up so we don't conflict.

AidanV committed:
Merge branch 'main' into vim-global-marks

AidanV commented:
@ConradIrwin I will pick up the database_id problem

AidanV committed:
marks from entity_id

ConradIrwin committed:
Re-instate writing to database

ConradIrwin commented:
Ok, pushed up some changes to actually write to the database; but seeing some odd behavior where it always resets marks to the length of the file on load (to investigate).

It also doesn't work for untitled buffers right now, I think we should make them hit the multi_buffer codepath until they are saved and then handle creation as a special form of rename (though still happy to cut that scope from v1).

I also think we should debounce database writes so we're not updating a given buffers marks more than a few times a second.

AidanV committed:
Untitled buffer marks

AidanV committed:
Renaming buffers

AidanV commented:
I added support for marking untitled buffers. 

I also added support for renaming files with their marks updating accordingly (including untitled buffers). 

Could you elaborate on what you meant by "seeing some odd behavior where it always resets marks to the length of the file on load"? If I am understanding this correctly then I am not seeing this issue.

The main things remaining in my eyes are to debounce database writes, and fix the issue of jumps occasionally traveling to the wrong location on the correct line (e.g.  `'` not jumping to the beginning of the line and `` ` `` not jumping to the mark position). This should be fixable by applying the same logic we currently have for local marks onto the other types of marks.

I noticed that you removed the harpoon like functionality for numbered marks. If this was intentional we should be sure to leave https://github.com/zed-industries/zed/issues/17353 open when this PR is finished.

ConradIrwin commented:
Nice, thanks!

The bug I saw was:
* Open a file (in this case LICENSE.MIT) set a mark for `c` on the start of "copyright" on the second paragraph, restart zed. Jump to mark 'c', it works, restart zed again, the mark is now pointing at the last position in the file.

I think we should merge without the numbered marks change – I'm not sure what the right behavior for them is, and I'd like to land the known work here first.

Hopefully backtick vs quote is just a missing if statement...

AidanV committed:
Fixed line jumps

ConradIrwin committed:
Merge branch 'main' into vim-global-marks

ConradIrwin committed:
Fix enabling/disabling vim

ConradIrwin committed:
Optimize database writing.

Each database write takes ~200µs. This takes
it from N per change (where N is the nubmer of marks below that position
in the file) to 1 per change.

ConradIrwin committed:
clipster

ConradIrwin commented:
Ok, I can no longer reproduce that bug; and I spent a bit time on the database writes (which seem fine - SQLite is fast!).

I also fixed a bad interaction when you disabled/re-enabled vim mode that was causing the tests to fail. 

Anything else you think we should look into before shipping this?

AidanV commented:
This looks ready to ship. Thank you for your help getting this done.

ConradIrwin committed:
Persist the . mark too

ConradIrwin committed:
Fix multibuffer marks (and merge the code-paths with untitled buffer marks).

ConradIrwin committed:
Small refactoring

ConradIrwin commented:
🥳 Alright merged!

For our next trick? `:marks` like `:registers`? or persist registers like marks?

I also noticed reviewing this we're missing `''` (#22398) and `'"` (not yet requested). We should also think about what to do with the numbered marks. Currently you can set them as global marks, but not sure if that solves the harpoon use-case well enough, do we need anything more automatic?

AidanV commented:
I have been looking forward to working on `:marks`, so I will probably get that done next. 

I don't think that the current global marks solution solves the harpoon use-case well enough unfortunately.

0x2CA commented:
>vim: Added persistence for global (and local) marks. When re-opening the same workspace your previous marks will be available.

Thank you for your work.

It works globally in nvim, you can jump without opening any file, it's similar to `:`, but in zed it doesn't work on EmptyPane

ConradIrwin commented:
@0x2CA I noticed that too :D. It's not trivial to fix as all our code to track state (like which mark you want to jump to) is per-editor instance, but it's probably not impossible either.

chrisarcand commented:
I was just thinking about this functionality like a week ago, awesome! Thanks for doing this.

ndvisme commented:
Hey, my goal is to implement a similar flow to the [harpoon nvim plugin](https://github.com/ThePrimeagen/harpoon).

I think this could be done using the [marks](https://github.com/zed-industries/zed/pull/25702)?
But I'm having a hard time to implement this.
I couldn't really find documentation on how to do this, how/what is currently supported in zed (looked in the command palate & in their docs website).

I was thinking something of the sort:
```
// keymap.json
[
  {
    "context": "Editor && vim_mode == normal && vim_operator == none && !VimWaiting",
    "bindings": {
      "space m h": [ "vim::SetMark", {"mark": "h"} ],
      "space m j": [ "vim::SetMark", {"mark": "j"} ],
      "space m k": [ "vim::SetMark", {"mark": "k"} ],
      "space m l": [ "vim::SetMark", {"mark": "l"} ]
    }
  },
  {
    "context": "Editor && vim_mode == normal",
    "bindings": {
      "ctrl-h": [ "vim::GoToMark", {"mark": "h"} ],
      "ctrl-j": [ "vim::GoToMark", {"mark": "j"} ],
      "ctrl-k": [ "vim::GoToMark", {"mark": "k"} ],
      "ctrl-l": [ "vim::GoToMark", {"mark": "l"} ]
    }
  }
]
```
Can someone point me in a certain direction to succeed in doing so?

Anyways, any help would be appreciated :)
Tnx in advance!

AidanV commented:
@ndvisme Below is one way to do what you are looking for. 

Uppercase marks are global marks which means that you can jump to them even when you are not currently in that buffer, which is likely what you are looking for. I know that harpoon does not jump to the location that was marked and instead just brings that buffer to the front without changing the cursor position. There currently is not a way to do that; however, I plan on implementing that feature soon. 

Feel free to add your thoughts on how this should be implemented [here](https://github.com/zed-industries/zed/issues/17353).
``` 
// keymap.json
[
  {
    "context": "Editor && vim_mode == normal && vim_operator == none && !VimWaiting",
    "bindings": {
      "space m h": ["workspace::SendKeystrokes", "m shift-h"],
      "space m j": ["workspace::SendKeystrokes", "m shift-j"],
      "space m k": ["workspace::SendKeystrokes", "m shift-k"],
      "space m l": ["workspace::SendKeystrokes", "m shift-l"]
    }
  },
  {
    "context": "Editor && vim_mode == normal",
    "bindings": {
      "ctrl-h": ["workspace::SendKeystrokes", "` shift-h"],
      "ctrl-j": ["workspace::SendKeystrokes", "` shift-j"],
      "ctrl-k": ["workspace::SendKeystrokes", "` shift-k"],
      "ctrl-l": ["workspace::SendKeystrokes", "` shift-l"]
    }
  }
]
```

I hope this helps!

ndvisme commented:
> @ndvisme Below is one way to do what you are looking for.
> 
> Uppercase marks are global marks which means that you can jump to them even when you are not currently in that buffer, which is likely what you are looking for. I know that harpoon does not jump to the location that was marked and instead just brings that buffer to the front without changing the cursor position. There currently is not a way to do that; however, I plan on implementing that feature soon.
> 
> Feel free to add your thoughts on how this should be implemented [here](https://github.com/zed-industries/zed/issues/17353).
> 
> ```
> // keymap.json
> [
>   {
>     "context": "Editor && vim_mode == normal && vim_operator == none && !VimWaiting",
>     "bindings": {
>       "space m h": ["workspace::SendKeystrokes", "m shift-h"],
>       "space m j": ["workspace::SendKeystrokes", "m shift-j"],
>       "space m k": ["workspace::SendKeystrokes", "m shift-k"],
>       "space m l": ["workspace::SendKeystrokes", "m shift-l"]
>     }
>   },
>   {
>     "context": "Editor && vim_mode == normal",
>     "bindings": {
>       "ctrl-h": ["workspace::SendKeystrokes", "` shift-h"],
>       "ctrl-j": ["workspace::SendKeystrokes", "` shift-j"],
>       "ctrl-k": ["workspace::SendKeystrokes", "` shift-k"],
>       "ctrl-l": ["workspace::SendKeystrokes", "` shift-l"]
>     }
>   }
> ]
> ```
> 
> I hope this helps!

Cool, this does the job for now! 
Thanks alot! 😄

tannersltwtr commented:
> @ndvisme Below is one way to do what you are looking for.
> 
> Uppercase marks are global marks which means that you can jump to them even when you are not currently in that buffer, which is likely what you are looking for. I know that harpoon does not jump to the location that was marked and instead just brings that buffer to the front without changing the cursor position. There currently is not a way to do that; however, I plan on implementing that feature soon.
> 
> Feel free to add your thoughts on how this should be implemented [here](https://github.com/zed-industries/zed/issues/17353).
> 
> ```
> // keymap.json
> [
>   {
>     "context": "Editor && vim_mode == normal && vim_operator == none && !VimWaiting",
>     "bindings": {
>       "space m h": ["workspace::SendKeystrokes", "m shift-h"],
>       "space m j": ["workspace::SendKeystrokes", "m shift-j"],
>       "space m k": ["workspace::SendKeystrokes", "m shift-k"],
>       "space m l": ["workspace::SendKeystrokes", "m shift-l"]
>     }
>   },
>   {
>     "context": "Editor && vim_mode == normal",
>     "bindings": {
>       "ctrl-h": ["workspace::SendKeystrokes", "` shift-h"],
>       "ctrl-j": ["workspace::SendKeystrokes", "` shift-j"],
>       "ctrl-k": ["workspace::SendKeystrokes", "` shift-k"],
>       "ctrl-l": ["workspace::SendKeystrokes", "` shift-l"]
>     }
>   }
> ]
> ```
> 
> I hope this helps!

Issue with this approach is if the file has already been marked with one of your 4 marks previously, you cant jump to the current iteration elsewhere because that file's mark doesn't get overwritten. We need a way to not make buffers have separate marks but rather project wide marks that will overwrite when reapplied. Unless of course this is possible and I am missing it?

AidanV commented:
@tannersltwtr Are you saying that with this config if you run `m shift-h` in file `A.txt` then navigate to file `B.txt` and do  `space m h` followed by `ctrl-h` it jumps back to file `A.txt`?

tannersltwtr commented:
@AidanV No it doesnt. Basically the issue I'm running into is if I have `A.txt` marked with `mH` previously, but I switched to `B.txt` and now want to set this file to `mH`, if I am ever currently viewing `A.txt` and want to swap to `B.txt` with ctrl-h, I can't because `A.txt` already contains that mark.  If we had a way of deleting marks this wouldn't be a huge issue though because I could delete them and reorganize my mark sets, but I couldn't find that option if it exists.

AidanV commented:
@tannersltwtr That is not the intended behavior, but I am not able to recreate that same problem. I can add deleting marks.

tannersltwtr commented:
@AidanV I can send you a video showing the issue for me. I'm thinking its actually a bug (or just inconsistency with vim) with overriding marks and them not setting properly.

AidanV commented:
@tannersltwtr #31140

tannersltwtr commented:
@AidanV Much appreciated!
